include/range/v3/at.hpp

```diff
@@ -14,35 +15,64 @@
 #ifndef RANGES_V3_AT_HPP
 #define RANGES_V3_AT_HPP
 
+#include <stdexcept>
 #include <range/v3/range_fwd.hpp>
 #include <range/v3/begin_end.hpp>
 #include <range/v3/range_traits.hpp>
 #include <range/v3/range_concepts.hpp>
-#include <range/v3/utility/iterator.hpp>
+#include <range/v3/distance.hpp>
 #include <range/v3/utility/static_const.hpp>
+#include <range/v3/index.hpp>
+
+RANGES_DISABLE_WARNINGS
+RANGES_DIAGNOSTIC_IGNORE_SIGN_CONVERSION
 
 namespace ranges
 {
     inline namespace v3
     {
+        /// Checked indexed range access.
+        ///
         /// \ingroup group-core
         struct at_fn
         {
             /// \return `begin(rng)[n]`
             template<typename Rng,
-                CONCEPT_REQUIRES_(RandomAccessRange<Rng>())>
+                CONCEPT_REQUIRES_(RandomAccessRange<Rng>() && SizedRange<Rng>())>
             RANGES_CXX14_CONSTEXPR
-            auto operator()(Rng &&rng, range_difference_type_t<Rng> n) const ->
+            auto operator()(Rng &&rng, range_size_type_t<Rng> n) const ->
```

Comment:
> If we use `difference_type` people calling `ranges::at(vec, vec.size() - 1)` will get warnings

Yup. In this case, they should just be using `back`, but the point stands. The question is how annoying and prescriptive we want to be in range-v3, given that the code range-v3 must play with does things a different way. I'm still undecided.